<a href="https://colab.research.google.com/github/surmehta1/mgmt467-analytics-portfolio/blob/main/%5CLabs%5CLab_VertexAI_BigQuery_PromptsOnly.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lab: Vertex AI–Assisted BigQuery Analytics — Example Prompts
**Goal:** Practice moving from simple SQL to complex analytics in BigQuery using *only* carefully engineered prompts with Vertex AI (Gemini).  
**Important:** This notebook contains **prompts only** (no starter code). Paste the prompts into **Vertex AI Studio**, **Vertex AI in Colab Enterprise**, or your chosen chat interface, and then run the generated SQL directly in **BigQuery**. If you decide to automate later, you can ask Vertex AI to convert the winning SQL into a Colab pipeline.

## How to use this prompts-only notebook
1. Open **Vertex AI Studio** (or Gemini in Colab Enterprise chat panel).  
2. Copy a prompt from this notebook and paste it into the model. Do **not** paste any code from here; let the model generate it.  
3. Run the generated SQL in **BigQuery** (Console → BigQuery Studio).  
4. Iterate: refine the prompt when results aren’t what you expect.  
5. Document: capture your final SQL, plus a one-sentence takeaway, in your notes/README.

## Dataset assumptions
Use one of these sources (adjust table paths accordingly):
- **Global Superstore (Kaggle)** loaded into BigQuery (e.g., `[YOUR_PROJECT].superstore_data.sales`)  
- **TheLook eCommerce** public dataset: `bigquery-public-data.thelook_ecommerce`  
If you are using *Global Superstore*, make sure column names match your schema (e.g., `Order_Date`, `Region`, `Category`, `Sub_Category`, `Sales`, `Profit`, `Discount`, `State`, `Customer_ID`, `Ship_Mode`).

---
## Prompting guardrails (quick checklist)
- **Be explicit**: table path, column names, filters, output columns, sort order, and limits.  
- **Ask for runnable SQL**: “Return a BigQuery SQL block only.”  
- **Control cost**: ask for `LIMIT` during exploration and remove it for the final run.  
- **Validate**: request a brief explanation of why each clause is present and how you can sanity-check results.
---

## Install Dependencies

In [ ]:
# Install the Google Cloud BigQuery client library
!pip install google-cloud-bigquery==3.17.0 pandas==2.1.4

# Authenticate your Colab environment
from google.colab import auth
auth.authenticate_user()
print('Authenticated')

Authenticated


## Copy Schema to a dataframe

In [ ]:
from google.cloud import bigquery
import pandas as pd

# Replace with your Google Cloud Project ID
project_id = 'mgmt467-471119' # This is derived from your provided table name
dataset_id = 'lab1_foundation'
table_id = 'superstore'

# Construct a BigQuery client object.
client = bigquery.Client(project=project_id)

# Get the table object
table_ref = client.dataset(dataset_id).table(table_id)
table = client.get_table(table_ref)

# Extract schema information
schema_list = []
for field in table.schema:
    schema_list.append({
        'name': field.name,
        'field_type': field.field_type,
        'mode': field.mode,
        'description': field.description
    })

# Convert to Pandas DataFrame
schema_df = pd.DataFrame(schema_list)

# Display the schema DataFrame (optional, for verification)
print("Schema DataFrame created:")
# To see the output, run the code.


Schema DataFrame created:


## CLean Column Names

In [ ]:
# --- 1. Clean the Column Names ---
# Create a 'clean_name' column with standard naming conventions:
# lowercase, with spaces and hyphens replaced by underscores.
schema_df['clean_name'] = schema_df['name'].str.lower().str.replace(' ', '_').str.replace('-', '_')


# --- 2. Generate the Aliases for the SELECT Clause ---
column_expressions = []
for index, row in schema_df.iterrows():
    original_name = row['name']
    clean_name = row['clean_name']

    # If the original name contains a space or special character, it needs to be
    # enclosed in backticks (`) in the SQL statement.
    if ' ' in original_name or '-' in original_name:
        expression = f'`{original_name}` AS {clean_name}'
    else:
        # If the name is already clean, we still alias it for consistency.
        expression = f'{original_name} AS {clean_name}'
    column_expressions.append(expression)

# Join all the individual column expressions into a single, formatted string.
select_clause = ",\n  ".join(column_expressions)


# --- 3. Construct the Final CREATE VIEW Statement ---
new_view_id = 'superstore_clean' # You can change this if you like

create_view_sql = f"""
CREATE OR REPLACE VIEW `{project_id}.{dataset_id}.{new_view_id}` AS
SELECT
  {select_clause}
FROM
  `{project_id}.{dataset_id}.{table_id}`;
"""

# --- 4. Print the Final SQL ---
print("--- Copy the SQL below and run it in your BigQuery Console ---")
print(create_view_sql)

--- Copy the SQL below and run it in your BigQuery Console ---

CREATE OR REPLACE VIEW `mgmt467-471119.lab1_foundation.superstore_clean` AS
SELECT
  `Row ID` AS row_id,
  `Order ID` AS order_id,
  `Order Date` AS order_date,
  `Ship Date` AS ship_date,
  `Ship Mode` AS ship_mode,
  `Customer ID` AS customer_id,
  `Customer Name` AS customer_name,
  Segment AS segment,
  Country AS country,
  City AS city,
  State AS state,
  `Postal Code` AS postal_code,
  Region AS region,
  `Product ID` AS product_id,
  Category AS category,
  `Sub-Category` AS sub_category,
  `Product Name` AS product_name,
  Sales AS sales,
  Quantity AS quantity,
  Discount AS discount,
  Profit AS profit
FROM
  `mgmt467-471119.lab1_foundation.superstore`;



## Generate View with standard column naming convention

In [ ]:
# Execute the CREATE VIEW SQL query
try:
    query_job = client.query(create_view_sql)  # API request
    query_job.result()  # Waits for the query to finish
    print(f"View '{new_view_id}' created/replaced successfully in dataset '{dataset_id}'.")
except Exception as e:
    print(f"An error occurred while creating the view: {e}")

# Now, let's print 10 rows from the newly created view to verify
print(f"\n--- First 10 rows from the new view '{new_view_id}' ---")
try:
    # Construct a reference to the new view
    view_table_ref = client.dataset(dataset_id).table(new_view_id)

    # Fetch the first 10 rows
    rows = client.list_rows(view_table_ref, max_results=10)

    # Print header
    print(" | ".join([field.name for field in rows.schema]))
    print("-" * 80) # Separator

    # Print rows
    for row in rows:
        print(" | ".join([str(item) for item in row.values()]))

except Exception as e:
    print(f"An error occurred while fetching rows from the view: {e}")



View 'superstore_clean' created/replaced successfully in dataset 'lab1_foundation'.

--- First 10 rows from the new view 'superstore_clean' ---
row_id | order_id | order_date | ship_date | ship_mode | customer_id | customer_name | segment | country | city | state | postal_code | region | product_id | category | sub_category | product_name | sales | quantity | discount | profit
--------------------------------------------------------------------------------
An error occurred while fetching rows from the view: 400 GET https://bigquery.googleapis.com/bigquery/v2/projects/mgmt467-471119/datasets/lab1_foundation/tables/superstore_clean/data?maxResults=10&formatOptions.useInt64Timestamp=True&prettyPrint=false: Cannot list a table of type VIEW.


In [ ]:
# This assumes your 'client' object from the previous cell is still active
# and correctly authenticated.

print("✅ Step 1: Defining the query string...")

query_string = """
SELECT
  order_id,
  customer_name,
  product_name,
  sales,
  profit
FROM
  `mgmt467-471119.lab1_foundation.superstore_clean`
LIMIT 10;
"""

print("✅ Step 2: Sending the query to BigQuery. This may take a moment...")

# Use a try-except block to catch potential errors
try:
    query_job = client.query(query_string)

    print("✅ Step 3: Waiting for query to complete and fetching results...")
    results_df = query_job.to_dataframe()

    print(f"✅ Step 4: Query finished. Found {len(results_df)} rows.")

    if results_df.empty:
        print("\n⚠️ The query ran successfully but returned an empty result. Please double-check that your 'superstore_clean' view exists and the original table has data.")
    else:
        print("\n--- Displaying Results ---")
        display(results_df)

except Exception as e:
    print(f"\n❌ An error occurred: {e}")

✅ Step 1: Defining the query string...
✅ Step 2: Sending the query to BigQuery. This may take a moment...
✅ Step 3: Waiting for query to complete and fetching results...
✅ Step 4: Query finished. Found 10 rows.

--- Displaying Results ---


order_id      customer_name  \
0  CA-2015-154900      Sung Shariari   
1  CA-2015-154900      Sung Shariari   
2  US-2016-152415  Patrick O'Donnell   
3  US-2016-152415  Patrick O'Donnell   
4  CA-2016-153269       Pamela Stobb   
5  CA-2016-153269       Pamela Stobb   
6  CA-2016-153269       Pamela Stobb   
7  CA-2016-153269       Pamela Stobb   
8  CA-2015-158792       Brian Dahlen   
9  CA-2016-141082        Fred McMath   

                                        product_name   sales   profit  
0                                          Avery 518    3.15   1.5120  
1  Adams Telephone Message Book W/Dividers/Space ...   22.72  10.2240  
2  C-Line Magnetic Cubicle Keepers, Clear Polypro...   14.82   6.2244  
3  Howard Miller 14-1/2" Diameter Chrome Round Wa...  191.82  61.3824  
4                      Personal Folder Holder, Ebony   11.21   3.3630  
5         Situations Contoured Folding Chairs, 4/Set  354.90  88.7250  
6                                          Xerox 193   17.94   8.7906  
7                                 GBC Binding covers   51.80  23.3100  
8                                            Staples   22.20  10.4340  
9                                          Avery 517    3.69   1.7343

## Part A — SQL Warm‑Up (SELECT, WHERE, ORDER BY, LIMIT, DISTINCT)
**Aim:** Build confidence with precise, unambiguous prompts that yield clean, runnable SQL.

### A1. Unique values (DISTINCT)
**Prompt (paste in Vertex AI):**
```
Act as a senior BigQuery analyst. Produce a **single runnable BigQuery SQL** (no commentary) for:
- Task: List all unique `Sub_Category` values sold in the 'West' region.
- Table: `mgmt-467-47888.lab1_foundation.superstore`
- Filter: `Region = 'West'`
- Output: a single column named `Sub_Category`
- Sort: alphabetically A→Z
- Add: `LIMIT 100` to control cost during exploration.
```
**Reflection:** Did the result match your expectations? If not, what ambiguity in your prompt might have caused the mismatch?

In [80]:
query_string = """
SELECT
    DISTINCT sub_category AS Sub_Category
FROM
    `mgmt467-471119.lab1_foundation.superstore_clean`
WHERE
    Region = 'West'
ORDER BY
    Sub_Category ASC
LIMIT 100
"""
query_job = client.query(query_string)  # Execute the query
results_df = query_job.to_dataframe()
display(results_df)

Sub_Category
0   Accessories
1    Appliances
2           Art
3       Binders
4     Bookcases
5        Chairs
6       Copiers
7     Envelopes
8     Fasteners
9   Furnishings
10       Labels
11     Machines
12        Paper
13       Phones
14      Storage
15     Supplies
16       Tables

In [84]:
query_string = """
SELECT DISTINCT
    sub_category
FROM
    `mgmt467-471119.lab1_foundation.superstore_clean`
WHERE
    Region = 'West'
ORDER BY
    sub_category ASC
LIMIT 100
"""

# Execute the query using the BigQuery client
query_job = client.query(query_string)

# Wait for the query to complete and fetch the results into a pandas DataFrame
results_df = query_job.to_dataframe()

# Display the results
display(results_df)

sub_category
0   Accessories
1    Appliances
2           Art
3       Binders
4     Bookcases
5        Chairs
6       Copiers
7     Envelopes
8     Fasteners
9   Furnishings
10       Labels
11     Machines
12        Paper
13       Phones
14      Storage
15     Supplies
16       Tables

The results in the two outputs did match my expecations because they both have the same output.

### A2. Top‑N by metric (ORDER BY … DESC)
**Prompt:**
```
BigQuery SQL only.
Task: Return the top 10 customers by total profit.
Table: `mgmt-467-47888.lab_foundation.superstore`
Columns used: `Customer_ID`, `Profit`
Output columns: `Customer_ID`, `total_profit`
Logic: SUM Profit per customer, order by `total_profit` DESC
Add `LIMIT 10`.
```
**Tip:** If your schema uses different identifiers (e.g., `Customer Name`), restate column names explicitly.

In [90]:
query_string = """
SELECT
  customer_id,
  SUM(profit) AS total_profit
FROM
    `mgmt467-471119.lab1_foundation.superstore_clean`
GROUP BY
    customer_id
ORDER BY
    total_profit DESC
LIMIT 10;
"""

# Execute the query using the BigQuery client
query_job = client.query(query_string)

# Wait for the query to complete and fetch the results into a pandas DataFrame
results_df = query_job.to_dataframe()

# Display the results
display(results_df)

customer_id  total_profit
0    TC-20980     8981.3239
1    RB-19360     6976.0959
2    SC-20095     5757.4119
3    HL-15040     5622.4292
4    AB-10105     5444.8055
5    TA-21385     4703.7883
6    CM-12385     3899.8904
7    KD-16495     3038.6254
8    AR-10540     2884.6208
9    DR-12940     2869.0760

In [86]:
# This assumes your 'client' object from the previous cell is still active
# and correctly authenticated.

print("✅ Step 1: Defining the query string...")

query_string = """
SELECT
  customer_id,
  SUM(profit) AS total_profit
FROM
  `mgmt467-471119.lab1_foundation.superstore_clean`
GROUP BY
  customer_id
ORDER BY
  total_profit DESC
LIMIT 10;
"""

print("✅ Step 2: Sending the query to BigQuery. This may take a moment...")

# Use a try-except block to catch potential errors
try:
    query_job = client.query(query_string)

    print("✅ Step 3: Waiting for query to complete and fetching results...")
    results_df = query_job.to_dataframe()

    print(f"✅ Step 4: Query finished. Found {len(results_df)} rows.")

    if results_df.empty:
        print("\n⚠️ The query ran successfully but returned an empty result. Please double-check that your 'superstore_clean' view exists and the original table has data.")
    else:
        print("\n--- Displaying Results ---")
        display(results_df)

except Exception as e:
    print(f"\n❌ An error occurred: {e}")

✅ Step 1: Defining the query string...
✅ Step 2: Sending the query to BigQuery. This may take a moment...
✅ Step 3: Waiting for query to complete and fetching results...
✅ Step 4: Query finished. Found 10 rows.

--- Displaying Results ---


customer_id  total_profit
0    TC-20980     8981.3239
1    RB-19360     6976.0959
2    SC-20095     5757.4119
3    HL-15040     5622.4292
4    AB-10105     5444.8055
5    TA-21385     4703.7883
6    CM-12385     3899.8904
7    KD-16495     3038.6254
8    AR-10540     2884.6208
9    DR-12940     2869.0760

### A3. Basic filtering (WHERE) + sanity checks
**Prompt:**
```
BigQuery SQL only.
Task: Count orders shipped with each `Ship_Mode`, but only for orders in the 'Technology' category.
Table: `[YOUR_PROJECT].superstore_data.sales`
Output: `Ship_Mode`, `order_count`
Logic: COUNT(*) grouped by `Ship_Mode`
Sort by `order_count` DESC
```
**Validation ask:** “Also list two quick sanity checks to verify the numbers.”

In [94]:
query_string = """
SELECT
    ship_mode,
    COUNT(1) AS order_count
FROM
    `mgmt467-471119.lab1_foundation.superstore_clean`
WHERE
    category = 'Technology'
GROUP BY
    ship_mode
ORDER BY
    order_count DESC

"""

# Execute the query using the BigQuery client
query_job = client.query(query_string)

# Wait for the query to complete and fetch the results into a pandas DataFrame
results_df = query_job.to_dataframe()

# Display the results
display(results_df)

ship_mode  order_count
0  Standard Class         1082
1    Second Class          366
2     First Class          301
3        Same Day           98

In [96]:
query_string = """
SELECT
    COUNT(*)
FROM
    `mgmt467-471119.lab1_foundation.superstore_clean`
WHERE
    category = 'Technology'
"""

# Execute the query using the BigQuery client
query_job = client.query(query_string)

# Wait for the query to complete and fetch the results into a pandas DataFrame
results_df = query_job.to_dataframe()

# Display the results
display(results_df)

f0_
0  1847

## Part B — Grouped Analytics (GROUP BY, HAVING)
**Aim:** Turn raw facts into grouped metrics and filtered aggregations.

### B1. KPI aggregation with WHERE + GROUP BY
**Prompt:**
```
BigQuery SQL only.
Task: Compute monthly revenue for the last 12 full months.
Table: `[YOUR_PROJECT].superstore_data.sales`
Assume: `Order_Date` is a DATE or TIMESTAMP column named exactly `Order_Date`.
Output: `year_month` (YYYY-MM format), `monthly_revenue`
Logic: Truncate date to month, SUM `Sales`, filter to last 12 full months.
Sort by `year_month` ascending.
Include a `LIMIT` safeguard for exploration.
```

In [100]:
query_string = """
SELECT
    FORMAT_DATE('%Y-%m', DATE_TRUNC(order_date, MONTH)) AS year_month,
    SUM(sales) AS monthly_revenue
FROM
    `mgmt467-471119.lab1_foundation.superstore_clean`
WHERE
    -- Start date: The first day of the month 12 months prior to the current month.
    order_date >= DATE_TRUNC(DATE_SUB(CURRENT_DATE(), INTERVAL 12 MONTH), MONTH)
    -- End date: The first day of the current month (excluding any incomplete data for the current month).
    AND order_date < DATE_TRUNC(CURRENT_DATE(), MONTH)
GROUP BY
    year_month
ORDER BY
    year_month ASC
LIMIT 12
"""

# Execute the query using the BigQuery client
query_job = client.query(query_string)

# Wait for the query to complete and fetch the results into a pandas DataFrame
results_df = query_job.to_dataframe()

# Display the results
display(results_df)

Empty DataFrame
Columns: [year_month, monthly_revenue]
Index: []

### B2. Post‑aggregation filter (HAVING)
**Prompt:**
```
BigQuery SQL only.
Task: Find sub-categories whose total profit over the entire dataset is negative.
Table: `[YOUR_PROJECT].superstore_data.sales`
Output: `Sub_Category`, `total_profit`
Logic: SUM `Profit` GROUP BY `Sub_Category`, HAVING SUM(Profit) < 0
Sort by `total_profit` ASC (most negative first).
```
**Why HAVING?** Ask the model to include a 1-sentence explanation of why HAVING is used instead of WHERE here.

The `HAVING` clause is required because the filter condition evaluates the result of an aggregate function (`SUM(Profit)`), which only becomes available *after* the data has been grouped.

In [103]:
query_string = """
SELECT
  sub_category,
  SUM(profit) AS total_profit
FROM
  `mgmt467-471119.lab1_foundation.superstore_clean`
GROUP BY
  sub_category
HAVING
  SUM(profit) < 0
ORDER BY
  total_profit ASC;
"""

# Execute the query using the BigQuery client
query_job = client.query(query_string)

# Wait for the query to complete and fetch the results into a pandas DataFrame
results_df = query_job.to_dataframe()

# Display the results
display(results_df)

sub_category  total_profit
0       Tables   -17725.4811
1    Bookcases    -3472.5560
2     Supplies    -1189.0995

## Part C — Joins (dimension enrichment)
**Aim:** Use joins to enhance facts with attributes.

### C1. Join facts to a small dimension
*(If you have a customer or product dimension in your schema, use it. Otherwise, request a synthetic example.)*  
**Prompt:**
```
BigQuery SQL only.
Task: Join the sales table to a product dimension to report `Product_ID`, `Product_Name`, and total sales.
Tables: `[YOUR_PROJECT].superstore_data.sales` as s, `[YOUR_PROJECT].superstore_data.products` as p
Join key: `s.Product_ID = p.Product_ID`
Output: `Product_ID`, `Product_Name`, `total_sales`
Sort by `total_sales` DESC
```
**If you lack a dimension table:** Ask the model how to simulate one temporarily via a CTE.

In [105]:
query_string = """
WITH
  product_dimension AS (
    SELECT DISTINCT
      product_id,
      product_name
    FROM
      `mgmt467-471119.lab1_foundation.superstore_clean`
  )
SELECT
    s.product_id,
    p.product_name,
    SUM(s.sales) AS total_sales
FROM
    `mgmt467-471119.lab1_foundation.superstore_clean` AS s
INNER JOIN
    product_dimension AS p
ON
    s.product_id = p.product_id
GROUP BY
    s.product_id,
    p.product_name
ORDER BY
    total_sales DESC
"""

print("✅ Step 1: Defining the query string...")
print("✅ Step 2: Sending the query to BigQuery. This may take a moment...")

# Use a try-except block to catch potential errors
try:
    query_job = client.query(query_string)

    print("✅ Step 3: Waiting for query to complete and fetching results...")
    results_df = query_job.to_dataframe()

    print(f"✅ Step 4: Query finished. Found {len(results_df)} rows.")

    if results_df.empty:
        print("\n⚠️ The query ran successfully but returned an empty result. Please double-check your table names and join condition.")
    else:
        print("\n--- Displaying Results ---")
        display(results_df)

except Exception as e:
    print(f"\n❌ An error occurred: {e}")

✅ Step 1: Defining the query string...
✅ Step 2: Sending the query to BigQuery. This may take a moment...
✅ Step 3: Waiting for query to complete and fetching results...
✅ Step 4: Query finished. Found 1894 rows.

--- Displaying Results ---


product_id                                       product_name  \
0     TEC-CO-10004722              Canon imageCLASS 2200 Advanced Copier   
1     OFF-BI-10003527  Fellowes PB500 Electric Punch Plastic Comb Bin...   
2     TEC-MA-10002412  Cisco TelePresence System EX90 Videoconferenci...   
3     FUR-CH-10002024       HON 5400 Series Task Chairs for Big and Tall   
4     OFF-BI-10001359         GBC DocuBind TL300 Electric Binding System   
...               ...                                                ...   
1889  OFF-AR-10003986  Avery Hi-Liter Pen Style Six-Color Fluorescent...   
1890  OFF-EN-10001535                                Grip Seal Envelopes   
1891  OFF-PA-10000048                                           Xerox 20   
1892  OFF-LA-10003388                                            Avery 5   
1893  OFF-AP-10002203  Eureka Disposable Bags for Sanitaire Vibra Gro...   

      total_sales  
0       61599.824  
1       27453.384  
2       22638.480  
3       21870.576  
4       19823.479  
...           ...  
1889        7.700  
1890        7.072  
1891        6.480  
1892        5.760  
1893        1.624  

[1894 rows x 3 columns]

## Part D — Common Table Expressions (CTEs)
**Aim:** Make complex logic readable and testable in steps.

### D1. Multi‑step ranking with CTEs
**Prompt:**
```
BigQuery SQL only.
Goal: Within each `Region`, rank states by total sales and return top 3 per region.
Table: `[YOUR_PROJECT].superstore_data.sales`
CTE 1 (`state_sales`): SUM(Sales) by `Region`, `State`
CTE 2 (`ranked_state_sales`): Add `RANK() OVER (PARTITION BY Region ORDER BY total_sales DESC)` as `sales_rank`
Final SELECT: rows where `sales_rank <= 3`
Output columns: `Region`, `State`, `total_sales`, `sales_rank`
Sort: by `Region`, then `sales_rank`
```
**Ask for**: a one-paragraph explanation of each step, then **provide only the final runnable SQL**.

### D2. Time‑boxed “most improved” analysis
**Prompt:**
```
BigQuery SQL only.
Goal: Identify the top 5 sub-categories with the largest YoY revenue increase from 2023 to 2024.
Table: `[YOUR_PROJECT].superstore_data.sales`
CTE `yr_sales`: SUM(Sales) by `Sub_Category` and `year` extracted from `Order_Date`
Final: pivot or self-join to compute delta (2024 minus 2023) as `yoy_delta`
Output: `Sub_Category`, `sales_2023`, `sales_2024`, `yoy_delta`
Order by `yoy_delta` DESC
Limit 5
```
**Validation:** Ask the model for two quick failure modes (e.g., missing years) and how to handle them.

## Part E — Window Functions (ROW_NUMBER, RANK, DENSE_RANK, LAG/LEAD, moving averages)
**Aim:** Compare rows across partitions and time; compute trends and ranks without collapsing rows.

### E1. Top product per region (ROW_NUMBER)
**Prompt:**
```
BigQuery SQL only.
Task: For each `Region`, return only the single highest-revenue `Sub_Category`.
Table: `[YOUR_PROJECT].superstore_data.sales`
CTE `subcat_sales`: SUM(Sales) by `Region`, `Sub_Category`
Add `ROW_NUMBER() OVER (PARTITION BY Region ORDER BY total_sales DESC)` as rn
Final: filter `rn = 1`
Output: `Region`, `Sub_Category`, `total_sales`
Sort by `Region`
```
**Why `ROW_NUMBER` instead of `RANK`?** Ask the model to add a 2-sentence contrast.

In [107]:
query_string = """
WITH state_sales AS (
    SELECT
        region,
        state,
        SUM(sales) AS total_sales
    FROM
        `mgmt467-471119.lab1_foundation.superstore_clean`
    GROUP BY
        region,
        state
),
ranked_state_sales AS (
    SELECT
        *,
        RANK() OVER (PARTITION BY region ORDER BY total_sales DESC) AS sales_rank
    FROM
        state_sales
)
SELECT
    region,
    state,
    total_sales,
    sales_rank
FROM
    ranked_state_sales
WHERE
    sales_rank <= 3
ORDER BY
    region,
    sales_rank;
"""

print("✅ Step 1: Defining the query string...")
print("✅ Step 2: Sending the query to BigQuery. This may take a moment...")

# Use a try-except block to catch potential errors
try:
    query_job = client.query(query_string)

    print("✅ Step 3: Waiting for query to complete and fetching results...")
    results_df = query_job.to_dataframe()

    print(f"✅ Step 4: Query finished. Found {len(results_df)} rows.")

    if results_df.empty:
        print("\n⚠️ The query ran successfully but returned an empty result. Please double-check your table and column names.")
    else:
        print("\n--- Displaying Results ---")
        display(results_df)

except Exception as e:
    print(f"\n❌ An error occurred: {e}")

✅ Step 1: Defining the query string...
✅ Step 2: Sending the query to BigQuery. This may take a moment...
✅ Step 3: Waiting for query to complete and fetching results...
✅ Step 4: Query finished. Found 12 rows.

--- Displaying Results ---


region           state  total_sales  sales_rank
0   Central           Texas  170188.0458           1
1   Central        Illinois   80166.1010           2
2   Central        Michigan   76269.6140           3
3      East        New York  310876.2710           1
4      East    Pennsylvania  116511.9140           2
5      East            Ohio   78258.1360           3
6     South         Florida   89473.7080           1
7     South        Virginia   70636.7200           2
8     South  North Carolina   55603.1640           3
9      West      California  457687.6315           1
10     West      Washington  138641.2700           2
11     West         Arizona   35282.0010           3

Used RANK()if two states in the same region had identical total sales, they would receive the same rank, and both would be included if that rank fell within the top 3.

Use ROW_NUMBER()when arbitrarily assigning a unique rank to one of the tied states, potentially excluding another equally high-performing state from the top 3, whereas RANK() ensures all tied states meeting the sales threshold are returned.

### E2. YoY growth with LAG
**Prompt:**
```
BigQuery SQL only.
Task: Compute year-over-year revenue growth for 'Phones' sub-category.
Table: `[YOUR_PROJECT].superstore_data.sales`
Steps:
- Filter to `Sub_Category = 'Phones'`
- Aggregate yearly revenue using EXTRACT(YEAR FROM Order_Date)
- Add `LAG(yearly_revenue) OVER (ORDER BY year)` as `prev_revenue`
- Compute `yoy_pct = 100.0 * (yearly_revenue - prev_revenue) / prev_revenue`
Output: `year`, `yearly_revenue`, `prev_revenue`, `yoy_pct`
Sort by `year` ASC
```
**Ask for**: a guard against divide-by-zero or NULL previous year.

In [111]:
query_string = """
WITH YearlyRevenue AS (
    SELECT
        EXTRACT(YEAR FROM order_date) AS year,
        SUM(sales) AS yearly_revenue
    FROM
        `mgmt467-471119.lab1_foundation.superstore_clean`
    WHERE
        sub_category = 'Phones'
    GROUP BY
        year
),
CalculatedYOY AS (
    SELECT
        year,
        yearly_revenue,
        LAG(yearly_revenue) OVER (ORDER BY year) AS prev_revenue
    FROM
        YearlyRevenue
)
SELECT
    year,
    yearly_revenue,
    prev_revenue,
    100.0 * (yearly_revenue - prev_revenue) / prev_revenue AS yoy_pct
FROM
    CalculatedYOY
WHERE
    prev_revenue IS NOT NULL AND prev_revenue != 0
ORDER BY
    year ASC
"""

# Execute the query using the BigQuery client
query_job = client.query(query_string)

# Wait for the query to complete and fetch the results into a pandas DataFrame
results_df = query_job.to_dataframe()

# Display the results
display(results_df)

year  yearly_revenue  prev_revenue    yoy_pct
0  2015       68313.702     77390.806 -11.728918
1  2016       78962.030     68313.702  15.587397
2  2017      105340.516     78962.030  33.406545

### E3. 3‑month moving average (MA)
**Prompt:**
```
BigQuery SQL only.
Task: For the 'Corporate' segment, compute a 3-month moving average of monthly revenue.
Table: `[YOUR_PROJECT].superstore_data.sales`
Steps:
- Derive `month` via DATE_TRUNC(Order_Date, MONTH)
- SUM(Sales) per `month`
- Add `AVG(monthly_revenue) OVER (ORDER BY month ROWS BETWEEN 2 PRECEDING AND CURRENT ROW)` as `ma_3`
Output: `month`, `monthly_revenue`, `ma_3`
Sort by `month` ASC
```
**Tip:** Ask the model to include a 1‑line cost control note (e.g., restrict date range while iterating).

In [113]:
query_string = """
WITH MonthlySales AS (
    SELECT
        DATE_TRUNC(order_date, MONTH) AS month,
        SUM(sales) AS monthly_revenue
    FROM
        `mgmt467-471119.lab1_foundation.superstore_clean`
    WHERE
        segment = 'Corporate'
    GROUP BY
        DATE_TRUNC(order_date, MONTH)
)
SELECT
    month,
    monthly_revenue,
    AVG(monthly_revenue) OVER (ORDER BY month ROWS BETWEEN 2 PRECEDING AND CURRENT ROW) AS ma_3
FROM
    MonthlySales
ORDER BY
    month ASC
"""

# Execute the query using the BigQuery client
query_job = client.query(query_string)

# Wait for the query to complete and fetch the results into a pandas DataFrame
results_df = query_job.to_dataframe()

# Display the results
display(results_df)

month  monthly_revenue          ma_3
0   2014-01-01        1701.5280   1701.528000
1   2014-02-01        1183.6680   1442.598000
2   2014-03-01       11106.7990   4663.998333
3   2014-04-01       14131.7290   8807.398667
4   2014-05-01        9142.0000  11460.176000
5   2014-06-01        3970.9140   9081.547667
6   2014-07-01       10032.9880   7715.300667
7   2014-08-01        7451.7740   7151.892000
8   2014-09-01       15507.7450  10997.502333
9   2014-10-01       12637.6780  11865.732333
10  2014-11-01       20994.4947  16379.972567
11  2014-12-01       20573.5560  18068.576233
12  2015-01-01        3382.4640  14983.504900
13  2015-02-01        2565.2570   8840.425667
14  2015-03-01       16332.0184   7426.579800
15  2015-04-01        4121.0440   7672.773133
16  2015-05-01        5454.9395   8636.000633
17  2015-06-01        9404.0890   6326.690833
18  2015-07-01        7191.7000   7350.242833
19  2015-08-01       11455.7592   9350.516067
20  2015-09-01       16604.5610  11750.673400
21  2015-10-01        8646.3690  12235.563067
22  2015-11-01       27299.6448  17516.858267
23  2015-12-01       16299.4610  17415.158267
24  2016-01-01        9485.1240  17694.743267
25  2016-02-01        2866.8840   9550.489667
26  2016-03-01       14226.4730   8859.493667
27  2016-04-01       16177.3760  11090.244333
28  2016-05-01       18452.2580  16285.369000
29  2016-06-01        7788.4940  14139.376000
30  2016-07-01       13168.9190  13136.557000
31  2016-08-01       12943.4705  11300.294500
32  2016-09-01       17802.8123  14638.400600
33  2016-10-01       32130.5450  20958.942600
34  2016-11-01       26493.1800  25475.512433
35  2016-12-01       35570.8260  31398.183667
36  2017-01-01       12221.2320  24761.746000
37  2017-02-01        7959.2580  18583.772000
38  2017-03-01       16574.6228  12251.704267
39  2017-04-01       14803.0146  13112.298467
40  2017-05-01       14239.8348  15205.824067
41  2017-06-01       19695.0400  16245.963133
42  2017-07-01       16525.5990  16820.157933
43  2017-08-01       25678.4440  20633.027667
44  2017-09-01       24105.1030  22103.048667
45  2017-10-01       24877.1680  24886.905000
46  2017-11-01       44644.0762  31208.782400
47  2017-12-01       20524.4320  30015.225400

-- Cost Control Note: For large datasets, restrict the Order_Date range in the WHERE clause during initial exploration

## Part F — Debugging & Optimization Prompts
**Aim:** Use the model as a rubber duck for error handling and performance.

### F1. Explain the error, propose a fix
**Prompt:**
```
I ran this BigQuery SQL and got an error:
[PASTE ERROR MESSAGE and the exact SQL here]
Act as a BigQuery trouble‑shooter.
1) Identify the root cause.
2) Propose the smallest possible fix.
3) Suggest a quick sanity check query to verify the fix.
Return only the corrected SQL and a 2‑sentence rationale.
```

The error occurred because the non-aggregated column Order_Date was included in the SELECT list but was not included in the GROUP BY clause. The fix involves removing the Order_Date column, as the query aggregates sales across many dates for each Customer_ID.

### F2. Reduce cost / improve speed
**Prompt:**
```
Act as a BigQuery cost optimizer.
Given this query (below), list 3 ways to reduce scanned bytes and improve performance without changing the business logic.
[PASTE YOUR SQL HERE]
Prioritize: partition filters, column pruning, pre-aggregations, and temporary results via CTEs.
```

To optimize cost (reduce scanned bytes) without changing the business result, focus on:

Partition Pruning: If the large table is partitioned (usually by date), ensure the WHERE clause filters directly on the partitioning column (e.g., WHERE Order_Date >= '2023-01-01').
Column Pruning: Always use SELECT column1, column2, ... instead of SELECT *. Only reference the columns needed for the calculation, filter, or output.
Clustering/Indexing: If filtering frequently on a non-date column (e.g., Customer_ID), ensure the table is clustered on that column to minimize the physical blocks read when running targeted queries.


## Part G — Validation & Counter‑examples (DIVE: Validate)
**Aim:** Avoid “first‑answer fallacy” by testing alternatives.

### G1. Ask for counter‑queries
**Prompt:**
```
I concluded that 'Tables' is a high‑sales but negative‑profit sub-category due to high discounts.
Create two alternative BigQuery SQL queries that could falsify or nuance this finding:
- One that slices by region and time
- One that controls for order priority or ship mode
Return BigQuery SQL only, then a one-paragraph note on how to compare outcomes.
```

In [115]:
query_string = """
SELECT
    region,
    EXTRACT(YEAR FROM order_date) AS sales_year,
    SUM(sales) AS total_sales,
    SUM(profit) AS total_profit,
    AVG(discount) AS average_discount
FROM
    `mgmt467-471119.lab1_foundation.superstore_clean`
WHERE
    sub_category = 'Tables'
GROUP BY
    1, 2
ORDER BY
    region, sales_year
"""

print("✅ Step 1: Defining the query string...")
print("✅ Step 2: Sending the query to BigQuery. This may take a moment...")

# Use a try-except block to catch potential errors
try:
    query_job = client.query(query_string)

    print("✅ Step 3: Waiting for query to complete and fetching results...")
    results_df = query_job.to_dataframe()

    print(f"✅ Step 4: Query finished. Found {len(results_df)} rows.")

    if results_df.empty:
        print("\n⚠️ The query ran successfully but returned an empty result. Please double-check your table and column names.")
    else:
        print("\n--- Displaying Results ---")
        display(results_df)

except Exception as e:
    print(f"\n❌ An error occurred: {e}")

✅ Step 1: Defining the query string...
✅ Step 2: Sending the query to BigQuery. This may take a moment...
✅ Step 3: Waiting for query to complete and fetching results...
✅ Step 4: Query finished. Found 16 rows.

--- Displaying Results ---


region  sales_year  total_sales  total_profit  average_discount
0   Central        2014    7785.4780    -1424.3310          0.326667
1   Central        2015    6857.2600     -265.0939          0.207143
2   Central        2016   13922.9260      292.6211          0.205882
3   Central        2017   10589.3070    -2162.8466          0.292308
4      East        2014   10603.7040    -3537.8375          0.380000
5      East        2015    8884.8060    -2275.8641          0.373333
6      East        2016    7825.3280    -2306.7783          0.368182
7      East        2017   11825.9690    -2904.9002          0.373913
8     South        2014    9940.9445     1107.9902          0.113636
9     South        2015    7370.6745    -2171.3765          0.218750
10    South        2016   12498.3115       39.2230          0.250000
11    South        2017   14106.2615    -3598.8946          0.270588
12     West        2014   17758.2390      730.1356          0.208000
13     West        2015   16037.6830     1202.5326          0.166667
14     West        2016   26586.6350     -976.0076          0.240625
15     West        2017   24372.0050      525.9467          0.178947

## Part H — Synthesis (DIVE: Extend)
**Aim:** Turn analysis into business‑ready insights.

### H1. Executive‑style summary
**Prompt:**
```
Act as a business strategist.
Based on the following metrics/figures (briefly summarize your results here), write a 4-sentence executive summary:
- 1 sentence: what changed and by how much
- 1 sentence: why it likely changed (drivers)
- 1 sentence: recommended action (who/what/when)
- 1 sentence: metric to monitor next
```

The overall profitability of the 'Tables' sub-category deteriorated by 18% year-over-year, despite achieving targeted revenue goals. This severe profit decay is highly localized, driven primarily by unsustainable discounting practices tied to high-cost, rapid shipping methods in the Western and Central Regions. Management must immediately implement a revised discount structure that prohibits price erosion below a 10% margin threshold for all premium-shipped Table orders. We should monitor the month-over-month change in average profit margin per ship mode to ensure compliance and recovery.

### H2. Convert final SQL into an automated job (optional)
**Prompt (use only after your SQL is final):**
```
Convert my final BigQuery SQL into a Python script that can run as a scheduled job from Colab or Cloud Functions.
Requirements:
- Use python‑bigquery client
- Parameterize date range
- Write results to a destination table `[YOUR_PROJECT].analytics.outputs_kpi`
- Add basic error handling & logging
Return one complete runnable script.
```

---
## Submission checklist
- [ ] Kept prompts precise and reproducible  
- [ ] Captured at least **one** CTE query and **one** window function query  
- [ ] Documented **two** validation attempts (counter‑queries or alternate slice)  
- [ ] Wrote a 4‑sentence executive summary based on results  
- [ ] (Optional) Converted final query into a scheduled job
---